concurrency: do not track timestamps for repl {exclusive,shared} locks

arulajmani opened the pull request:
Previously, the timestamp tracking for replicated locks would be updated
whenever a lock was acquired/rediscovered, regardless of the strength.
This wasn't great -- the timestamp of a replicated lock is only relavent
if it's strength is lock.Intent. Worse yet, because we were blindly
forwarding the timestamp tracking, non-locking readers could end up in
infinite lock discovery loops.

This patch fixes such issues by only tracking/forwarding timestamps
for intents when it comes to replicated locks.

Closes https://github.com/cockroachdb/cockroach/issues/109673

Release note: None

arulajmani committed:
concurrency: split getLockHolder into 2 functions

Burn down a TODO.

Epic: none

Release note: None

arulajmani committed:
concurrency: do not track timestamps for repl {exclusive,shared} locks

Previously, the timestamp tracking for replicated locks would be updated
whenever a lock was acquired/rediscovered, regardless of the strength.
This wasn't great -- the timestamp of a replicated lock is only relavent
if it's strength is lock.Intent. Worse yet, because we were blindly
forwarding the timestamp tracking, non-locking readers could end up in
infinite lock discovery loops.

This patch fixes such issues by only tracking/forwarding timestamps
for intents when it comes to replicated locks.

Closes https://github.com/cockroachdb/cockroach/issues/109673

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/110151)
<!-- Reviewable:end -->

nvb reviewed (commented):
> Reviewed 2 of 2 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110151)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 1315 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110151#-Ndl1BY730HjDYc6h_HZ:-Ndl1BY730HjDYc6h_H_:b-pkd1as) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5a3f46c238b74abc51466ec6ed387a9aa3ce84e/pkg/kv/kvserver/concurrency/lock_table.go#L1315)):*
> ```Go
> //
> // REQUIRES: kl.mu to be locked.
> // TODO(XXX): having trouble with the name here.
> ```

We could name it something like `mutationTS` or `writeTS`. The fundamental reason why non-locking readers need to block on intents and not other locks is because there's a pending mutation protected by the intent and the non-locking reader may need to observe that value.
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 1330 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110151#-Ndl1kbA1hDwppBGVzLJ:-Ndl1kbA1hDwppBGVzLK:bvy7cqy) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5a3f46c238b74abc51466ec6ed387a9aa3ce84e/pkg/kv/kvserver/concurrency/lock_table.go#L1330)):*
> ```Go
> 		// If there's both a write intent and an unreplicated exclusive lock, we want
> 		// to prefer the lower of the two timestamps, since the lower timestamp
> 		// blocks more non-locking readers.
> ```

What if `kv.lock.exclusive_locks_block_non_locking_reads.enabled` is set to false? I think it might be easier to just say that if an unreplicated exclusive lock and an intent are both present, defer to the intent timestamp. There's no need to block on an earlier exclusive lock if an intent is stored at a higher timestamp.
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 1727 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110151#-Ndl2_bL36huj8HzkT5v:-Ndl2_bL36huj8HzkT5w:b-xrztby) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5a3f46c238b74abc51466ec6ed387a9aa3ce84e/pkg/kv/kvserver/concurrency/lock_table.go#L1727)):*
> ```Go
> 			// formatting. The lowest timestamp contends with more transactions
> 			// (assuming the strength dictates as such).
> 			var ts hlc.Timestamp
> ```

Should this just call `nonLockingReadersBlockingTS()`?


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110151)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 1727 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110151#-Ndl2_bL36huj8HzkT5v:-Ndl3uW4-eGQEqRQWJYb:b-yqrvxa) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5a3f46c238b74abc51466ec6ed387a9aa3ce84e/pkg/kv/kvserver/concurrency/lock_table.go#L1727)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should this just call `nonLockingReadersBlockingTS()`?
</blockquote></details>

It could, but then all the test files that acquire a shared lock will churn with s/txn_ts/hlc.MaxTimestamp/g. Another option is to only print the timestamp if the lock is held with strengths >= lock.Exclusive and omit it otherwise. How do you feel about the second option?


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110151)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 1727 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110151#-Ndl2_bL36huj8HzkT5v:-Ndl5e4E2tq9bdIaUMgs:b-kjbzbl) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5a3f46c238b74abc51466ec6ed387a9aa3ce84e/pkg/kv/kvserver/concurrency/lock_table.go#L1727)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

It could, but then all the test files that acquire a shared lock will churn with s/txn_ts/hlc.MaxTimestamp/g. Another option is to only print the timestamp if the lock is held with strengths >= lock.Exclusive and omit it otherwise. How do you feel about the second option?
</blockquote></details>

So only print the timestamp if it's not MaxTimestamp? That SGTM.


<!-- Sent from Reviewable.io -->

arulajmani committed:
concurrency: omit timestamp for shared lock formatting

Timestamp information is irrelevant to shared locks; omit it when
formatting them.

Epic: none

Release note: None

arulajmani committed:
concurrency: handle unrepl exclusive lock/repl intent interactions

This patch defines the behavior when both an exclusive unreplicated
lock and an intent are present on a key. In such cases, we use the
intent's timestamp for conflict resolution, regardless of the exclusive
lock's timestamp. The meaningful case is where the exclusive lock is
held at a lower timestamp than the intent -- in such cases, non-locking
reads that lie between [exclusiveLockTS, intentTS) are allowed to
proceed without blocking.

Epic: none

Release note: None

arulajmani reviewed (commented):
@nvanbenschoten I've added 2 new commits to this patch based on your feedback -- r3 is ~churn and r4 defines the interaction when both an intent and exclusive locks exist on a key. PTAL. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110151#-:-NdlU3cp2HhlC1Sle-rP:b4gk67e)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 1315 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110151#-Ndl1BY730HjDYc6h_HZ:-NdlMoBd84jAZ28Vb5Bn:b-j1tpt5) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5a3f46c238b74abc51466ec6ed387a9aa3ce84e/pkg/kv/kvserver/concurrency/lock_table.go#L1315)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We could name it something like `mutationTS` or `writeTS`. The fundamental reason why non-locking readers need to block on intents and not other locks is because there's a pending mutation protected by the intent and the non-locking reader may need to observe that value.
</blockquote></details>

Switched to `writeTS`. 
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 1330 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110151#-Ndl1kbA1hDwppBGVzLJ:-NdlTkZx69X4K6kxmeAS:b-hlb4t2) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5a3f46c238b74abc51466ec6ed387a9aa3ce84e/pkg/kv/kvserver/concurrency/lock_table.go#L1330)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

What if `kv.lock.exclusive_locks_block_non_locking_reads.enabled` is set to false? I think it might be easier to just say that if an unreplicated exclusive lock and an intent are both present, defer to the intent timestamp. There's no need to block on an earlier exclusive lock if an intent is stored at a higher timestamp.
</blockquote></details>

That's a good call. To make things work correctly, we also need to use this function in `getLockMode` (instead of reaching into `{,un}replicatedLockHolderInfo directly). I've made that change and added some testing on the non-locking scan phase as well -- it's in its own commit, r4.
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 1727 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110151#-Ndl2_bL36huj8HzkT5v:-NdlMqEc2Ez7VVEYTN6s:b2hd1v0) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5a3f46c238b74abc51466ec6ed387a9aa3ce84e/pkg/kv/kvserver/concurrency/lock_table.go#L1727)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

So only print the timestamp if it's not MaxTimestamp? That SGTM.
</blockquote></details>

Tacked on third commit for this. It's mostly test churn.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r5, 2 of 2 files at r6, 6 of 6 files at r7, 4 of 4 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110151#-:-Ndq5pILCzFZRoiIKW8u:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 1311 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/110151#-Ndq4XpkE2Mtywj7JT7g:-Ndq4XpkE2Mtywj7JT7h:b-bkbzrf) ([raw file](https://github.com/cockroachdb/cockroach/blob/c506bb0c43b2bf5d3ab9da21bc7b329d6d3a9c91/pkg/kv/kvserver/concurrency/lock_table.go#L1311)):*
> ```Go
> 
> // writeTS returns a timestamp at which a provisional write has been
> // performed[*]. Any non-locking readers above this timestamp should block on
> ```

s/above/at or above/
___
*[`pkg/kv/kvserver/concurrency/lock_table_test.go` line 1928 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/110151#-Ndq51Li7C0SQw__mmmr:-Ndq51Li7C0SQw__mmms:brnm3rv) ([raw file](https://github.com/cockroachdb/cockroach/blob/93931a018fff5180fe35db9dfc8aa91eb261e38f/pkg/kv/kvserver/concurrency/lock_table_test.go#L1928)):*
> ```Go
> 	require.EqualValues(t,
> 		" lock: ‹\"KEY\"›\n"+
> 			"  holders: txn: 6ba7b810-9dad-11d1-80b4-00c04fd430c8 epoch: 0, iso: Serializable, ts: 0.000000123,7, info: unrepl [(str: Shared seq: 3)]\n"+
> ```

Should we adjust this formatting test to acquire an Intent so that we don't lose this test coverage?


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
TFTR!

bors r=nvanbenschoten

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110151#-:-NdqKdQCE8kDyAmrMO2_:b-l0r5kv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[`pkg/kv/kvserver/concurrency/lock_table_test.go` line 1928 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/110151#-Ndq51Li7C0SQw__mmmr:-NdqK8hS1oThH9CGI1ni:brku4dd) ([raw file](https://github.com/cockroachdb/cockroach/blob/93931a018fff5180fe35db9dfc8aa91eb261e38f/pkg/kv/kvserver/concurrency/lock_table_test.go#L1928)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we adjust this formatting test to acquire an Intent so that we don't lose this test coverage?
</blockquote></details>

I think that should be covered by `TestLockStateSafeFormat`, which acquires with both lock.Shared and lock.Intent. This one is just about making sure multiple lock holders are formatted correctly -- I'm leaning towards keeping this as is, because having two shared locks and an intent on a key is unrealistic.


<!-- Sent from Reviewable.io -->
